### PR TITLE
Exercise solution

### PR DESCRIPTION
**Pull request by clicktravel-martin**
I haven't done anything about the following, as they're beyond the scope of a throwaway exercise - we undoubtedly have standards/patterns/libraries for them at TK, so it didn't seem worth spending time inventing something:
- i18n / L10n
- a11y
- Form validation beyond native browser behaviour
- REST error handling and normalisation of error messages
- Linting/code style beyond what you get by default with `create-react-app`
- TypeScript
- Responsive/adaptive design
- Styling that looks nice 😄 - the HTML is semantically correct though

### CODE REVIEW

File: src/add-recipe/AddRecipe.js [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  }
+
+  useEffect(() => {
+    if (recipeToSave) {
+      (async () => {
+        await recipeListEndpointAdapter.post({ recipe: recipeToSave });
+        setStep(Step.SAVED);
+      })();
+    }
+  }, [recipeToSave]);
```
Comment (clicktravel-martin) [inline src/add-recipe/AddRecipe.js:31]: Everywhere that HTTP requests happen after user interaction, I've done it declaratively by having the side-effect happen inside a `useEffect` triggered by a change in state. My understanding is that this is the preferred React way of doing things, but I'm also aware that in many cases it works fine if you skip `useEffect` entirely and just call endpoints etc. imperatively in event handlers - the web is full of examples of this. Does TK have a preferred position on this?

File: src/App.css [hunk 2]
```diff
@@ -0,0 +1,38 @@
+.App {
```
Comment (ivan-navarro-75) [inline src/App.css:1]: I think this css file can be safely removed

File: src/add-recipe/AddRecipe.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  const [recipeToSave, setRecipeToSave] = useState();
+
+  function saveRecipe(recipe) {
+    setStep(Step.SAVING);
+    setRecipeToSave(recipe);
+  }
+
+  useEffect(() => {
+    if (recipeToSave) {
+      (async () => {
```
Comment (ivan-navarro-75) [inline src/add-recipe/AddRecipe.js:26]: I'd declare the function and the call it separately, mainly for readability

File: src/add-recipe/AddRecipe.js (hunk 1 above)
Comment (ivan-navarro-75) [reply src/add-recipe/AddRecipe.js:31]: what you're doing here it is a kind of optimistic update, you partially change the change the component state and then sync this state with the backend side. 
In your case I wouldn't update the state, triggering the rerender and then capturing the change with the useEffect
I think everything is easier if you something similar to this, and it is usually the pattern use in TK.

```
const saveRecipe = async(recipe) => {
    try{
       setStep(Step.SAVING);
       await recipeListEndpointAdapter.post({ recipe: recipeToSave });
       setStep(Step.SAVED);
   }
   catch(e){
      setError(e);
   }
}
```

File: src/add-recipe/AddRecipe.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+        await recipeListEndpointAdapter.post({ recipe: recipeToSave });
+        setStep(Step.SAVED);
+      })();
+    }
+  }, [recipeToSave]);
+
+  return (
+    <div>
+      <H1>Add a recipe</H1>
```
Comment (ivan-navarro-75) [inline src/add-recipe/AddRecipe.js:35]: I am missing some error control here

File: src/add-recipe/AddRecipe.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+import { H1, P } from '../shared/styled';
+
+const Step = {
+  EDITING: 'EDITING',
+  SAVING: 'SAVING',
+  SAVED: 'SAVED',
+};
+
+const AddRecipe = () => {
+  const [step, setStep] = useState(Step.EDITING);
```
Comment (ivan-navarro-75) [inline src/add-recipe/AddRecipe.js:16]: love this pattern

File: src/add-recipe/AddRecipe.test.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    act(() => {
+      saveRecipeCallback({
+        name: 'My recipe',
+        description: 'Delicious',
+        ingredients: [
+          { name: 'Sausage' },
+          { name: 'Apricot' },
+        ],
+      });
+    });
```
Comment (ivan-navarro-75) [inline src/add-recipe/AddRecipe.test.js:37]: wow! I think it is hard to follow what are you doing here, you are mocking the whole form and then recover callback from it to call it with your own params, I didn't even know that was possible. 

Maybe if the generation of `RecipeForm` is super heavy there is a reason to mock it, but if it is not the case, why not just use the real component?

File: src/manage-recipe/ManageRecipe.js [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    setRecipeToUpdate(recipe);
+  }
+
+  function deleteRecipe() {
+    setStep(Step.DELETING);
+  }
+
+  useEffect(() => {
+    const { recipeId } = match.params;
+    (async () => {
```
Comment (ivan-navarro-75) [inline src/manage-recipe/ManageRecipe.js:37]: again, I think this pattern is adding a complexity that it is not needed

File: src/manage-recipe/ManageRecipe.js [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+      const recipe = await recipeEndpointAdapter.get({ recipeId });
+      setRecipe(recipe);
+      setStep(Step.EDITING);
+    })();
+  }, [match.params]); // match.params doesn't change during the page's life, so this only runs once on load
+
+  useEffect(() => {
+    const { recipeId } = match.params;
```
Comment (ivan-navarro-75) [inline src/manage-recipe/ManageRecipe.js:45]: same, I think a single function for each callback, which updates the state, calls the end-point and sets again the state again regarding the previous response is easier to follow than involve our friends useEffects here

File: src/manage-recipe/ManageRecipe.test.js [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+        { name: 'Egg' },
+        { name: 'Sausage meat' },
+        { name: 'Breadcrumbs' },
+      ],
+    });
+    recipeEndpointAdapter.patch.mockResolvedValue({ id: 'DUMMY_VALUE' });
+    render(<ManageRecipe match={{ params: { recipeId: '12345' } }}/>, { wrapper: MemoryRouter });
+    await waitForElementToBeRemoved(() => screen.getByText('Please wait...'));
+
+    const { saveRecipeCallback } = RecipeForm.mock.calls[0][0];
```
Comment (ivan-navarro-75) [inline src/manage-recipe/ManageRecipe.test.js:69]: again, those tests are too much coupled with your implementation details, it is true that you're isolating this test to the rest of the world but we try to test our code as if there is a user playing with it, and accessing to the callbacks is not something a regular user does

File: src/shared/recipeEndpointAdapter.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
+import axios from 'axios';
+
+const ENDPOINT_URL = '/recipes/{recipeId}/';
+
+const get = async ({ recipeId }) => {
+  const response = await axios.get(ENDPOINT_URL.replace('{recipeId}', recipeId), {
```
Comment (ivan-navarro-75) [inline src/shared/recipeEndpointAdapter.js:6]: with [template literals](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) you can achieve the same easily

File: src/shared/RecipeForm.js [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    };
+    setIngredients(newIngredients);
+  }
+
+  function addIngredient() {
+    setIngredients([...ingredients, { name: '' }]);
+  }
+
+  function removeIngredient(index) {
+    setIngredients([...ingredients].filter((ingredient, i) => i !== index));
```
Comment (ivan-navarro-75) [inline src/shared/RecipeForm.js:40]: filter always returns a new array, so it is not need to copy the original before filtering

File: src/shared/RecipeForm.test.js [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+        { name: 'Sausage meat' },
+        { name: 'Breadcrumbs' },
+      ],
+    };
+    render(<RecipeForm saveRecipeCallback={NOOP} recipe={recipe}/>, { wrapper: MemoryRouter });
+
+    userEvent.click(within(screen.getAllByRole('listitem')[1]).getByRole('button', { name: 'Remove' }));
+
+    const ingredients = within(screen.getByRole('list')).getAllByRole('listitem');
+    expect(ingredients.length).toBe(2);
```
Comment (ivan-navarro-75) [inline src/shared/RecipeForm.test.js:104]: it would be great to check that before the removal we had 3

**Review by ivan-navarro-75 — COMMENTED** [review]
nice job! please check my comments and if you need further explanation don't hesitate to contact me directly
